### PR TITLE
Dev f-strings

### PR DESCRIPTION
**Pull request by madeddy**
Port to f-string formating
- The files magic.py and codegen.py where because py2 compat omitted.

### CODE REVIEW

**Comment by CensoredUsername** [comment]
Ah, minimize/codegen probably don't support f-strings yet as those didn't exist when they were created. I'll take a look.

**Comment by madeddy** [comment]
Well it was to expect. The only test i forgot to run uncovering the bugs, is the icing on the cake.

> Ah, minimize/codegen probably don't support f-strings yet as those didn't exist when they were created. I'll take a look.

Thanks a lot!
I expected its something in the injector files, as unrpyc worked in all tests, but i could not find the real source.

**Comment by CensoredUsername** [comment]
I just started working on this, and it is surprisingly nontrivial (until python 3.12 there's a _lot_ of weird edge cases around f-string generation relating to reusing quote symbols). Might take me a bit to have the time to work on this, I need to figure out a way to actually handle this properly.

**Comment by madeddy** [comment]
I had also a dilemma with this quotes and forbidden bslash. To get on a uniform usage of the quotes i put all f-strings in double quotes and everything inside in single q. After like 200 i did first tests and had to find Ren'py couldn't stomach this, because i overlooked the inner single quotes are written out with the strings. Naturally i had to reverse all.

I am glad they figured this in py3.12 out and we have more usage freedom.

> Might take me a bit to have the time to work on this

No prob at all. To have f-string is imo nice to have finally, but for now the app works. Also i didn't notice till the checks fail, because i compiled all the time (repeat use of command) with option -r. So without minimization it runs so far.

**Comment by CensoredUsername** [comment]
Aight, I think I've handled this properly on dev now. Unfortunately github is a bit dumb and I literally cannot rerun tests with changes on the target branch sorted out, but it all seems to work locally.

I'll review the actual pull tomorrow likely.

**Comment by madeddy** [comment]
> I think I've handled this properly on dev now.
> I'll review the actual pull tomorrow likely.

Oh, good. I expected it not so fast. 👍🏻 

> cannot rerun tests

I think a different workflow config is needed to get the rerun button or other yaml options to trigger this.:
[re-running-workflows-and-jobs](https://docs.github.com/en/actions/managing-workflow-runs/re-running-workflows-and-jobs)
[events-that-trigger-workflows#check_run](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#check_run)

File: decompiler/__init__.py [hunk 1]
```diff
@@ -261,21 +261,21 @@ def print_with(self, ast):
             # Sanity check. check if there's a matching with statement two nodes further
             if not(isinstance(self.block[self.index + 2], renpy.ast.With) and
                    self.block[self.index + 2].expr == ast.paired):
-                raise Exception("Unmatched paired with {0} != {1}".format(
-                                repr(self.paired_with), repr(ast.expr)))
+                raise Exception(
+                    f'Unmatched paired with {repr(self.paired_with)} != {repr(ast.expr)}')
```
Comment (CensoredUsername) [inline decompiler/__init__.py:265]: You can use the !r conversion here for both items, as in `{self.paired_with!r}`

File: decompiler/translate.py [hunk 2]
```diff
@@ -69,7 +69,7 @@ def create_translate(self, block):
             elif isinstance(i, renpy.ast.UserStatement):
                 code = i.line
             else:
-                raise Exception("Don't know how to get canonical code for a %s" % str(type(i)))
+                raise Exception(f'Don\'t know how to get canonical code for a {str(type(i))}')
```
Comment (CensoredUsername) [inline decompiler/translate.py:72]: Can use the !s conversion here

File: decompiler/util.py [hunk 3]
```diff
@@ -189,11 +189,11 @@ def print_debug(self, message):
     def write_failure(self, message):
         self.print_debug(message)
         self.indent()
-        self.write("pass # <<<COULD NOT DECOMPILE: %s>>>" % message)
+        self.write(f'pass # <<<COULD NOT DECOMPILE: {message}>>>')
 
     def print_unknown(self, ast):
         # If we encounter a placeholder note, print a warning and insert a placeholder
-        self.write_failure("Unknown AST node: %s" % str(type(ast)))
+        self.write_failure(f'Unknown AST node: {str(type(ast))}')
```
Comment (CensoredUsername) [inline decompiler/util.py:196]: also `!s` instead of the `str` call

**Review by CensoredUsername — COMMENTED** [review]
I added some comments on where you can use `!r` and `!s` formatting conversions instead of explicit str/repr calls. But it looks good regardless.

**Comment by madeddy** [comment]
Hehe, ok. I wanted to ask a few weeks ago if you prefer them changed or not. Simply forgot about it. Going to add the relating changes later.

File: decompiler/__init__.py (hunk 1 above)
Comment (madeddy) [reply decompiler/__init__.py:265]: Done

File: decompiler/translate.py (hunk 2 above)
Comment (madeddy) [reply decompiler/translate.py:72]: Done

File: decompiler/util.py (hunk 3 above)
Comment (madeddy) [reply decompiler/util.py:196]: Done

**Comment by CensoredUsername** [comment]
Aight, thanks!
